Rules/AvoidTrailingWhitespace.cs

```diff
@@ -36,52 +36,68 @@ public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
 
             var diagnosticRecords = new List<DiagnosticRecord>();
 
-            string[] lines = Regex.Split(ast.Extent.Text, @"\r?\n");
+            //string[] lines = Regex.Split(ast.Extent.Text, @"\r?\n");
+            string[] lines = ast.Extent.Text.Split(new[] { "\r\n", "\r", "\n" }, StringSplitOptions.None);
 
             for (int lineNumber = 0; lineNumber < lines.Length; lineNumber++)
             {
                 var line = lines[lineNumber];
 
-                var match = Regex.Match(line, @"\s+$");
-                if (match.Success)
+                if (line.Length == 0)
                 {
-                    var startLine = lineNumber + 1;
-                    var endLine = startLine;
-                    var startColumn = match.Index + 1;
-                    var endColumn = startColumn + match.Length;
-
-                    var violationExtent = new ScriptExtent(
-                        new ScriptPosition(
-                            ast.Extent.File,
-                            startLine,
-                            startColumn,
-                            line
-                        ),
-                        new ScriptPosition(
-                            ast.Extent.File,
-                            endLine,
-                            endColumn,
-                            line
-                        ));
-
-                    var suggestedCorrections = new List<CorrectionExtent>();
-                    suggestedCorrections.Add(new CorrectionExtent(
-                                violationExtent,
-                                string.Empty,
-                                ast.Extent.File
-                            ));
+                    continue;
+                }
+                if (line[line.Length - 1] != ' ' &&
```

Comment:
@rjmholt  because of
- readablity
- perfomance
- covering the variety of unicode chars? from the docs [here](https://docs.microsoft.com/en-us/dotnet/api/system.char.iswhitespace?view=netcore-3.1#remarks), it would probably be good but what about the [UnicodeCategory.LineSeparator](https://docs.microsoft.com/en-us/dotnet/api/system.globalization.unicodecategory?view=netcore-3.1#System_Globalization_UnicodeCategory_LineSeparator) char? I don't have much Unicode experience to make a judgement call here tbh
![image](https://user-images.githubusercontent.com/9250262/80420515-ca7e3580-88d2-11ea-8a5c-5bcf59a55d84.png)

Reply:
My thinking here is actually just that [PowerShell uses that API to see whitespace](https://github.com/PowerShell/PowerShell/blob/1c4ecb6ec5503e0693d58f31278dc44140fc30c7/src/System.Management.Automation/engine/parser/tokenizer.cs#L964-L976).

Given how we split the string already, it's possibly dangerous to go by unicode whitespace, but possibly not...

I suspect that really this won't make much difference; leaving non-ASCII whitespace at the ends of lines isn't something I can imagine being an issue for anyone really.

Reply:
Ok, so that sounds more like a tendency to use `IsWhiteSpace`? I'd be OK with that, you are right that the impact is probably quite low, especially since this rules is not enabled by default for vs-code users.